### ISSUE TITLE
Kaiser windowing in FftSharp

### ISSUE THREAD

**Issue by ADD-eNavarro** [swharden/FftSharp#35]
Hi!

Great work here! I am looking forward to use your FftSharp library, but I would need to have a Kaiser windowing implementation working. Do you plan on implementing that? Or else, do you know some other library I could use tith Kaiser windowing implemented?

**Comment by swharden** [comment]
For example the Hanning window is just:

https://github.com/swharden/FftSharp/blob/6f51c470f2c7327180070c140c93f890e0630689/src/FftSharp/Window.cs#L67-L75

**Comment by ADD-eNavarro** [comment]
Absolutelly!

[Here](https://tomroelandts.com/articles/how-to-create-a-configurable-filter-using-a-kaiser-window)'s an implementation in Python, may be helpful if you're going to implement Kaiser window in your awesome tool.

**Comment by swharden** [comment]
Note to self: https://numpy.org/doc/stable/reference/generated/numpy.kaiser.html

**Comment by swharden** [comment]
So this turned out to be way harder than I expected...

Calculating the Kaiser window is easy. It's a modified zero-order Bessel function. In python it's just this ([source](https://tomroelandts.com/articles/how-to-create-a-configurable-filter-using-a-kaiser-window)):

```py
w = np.i0(beta * np.sqrt(1 - (2 * n / (N - 1) - 1) ** 2)) / np.i0(beta)
```

The hard part is figuring out what that `np.i0()` is doing. It looks like it's generating a "Modified Bessel function of the first kind, order 0." https://numpy.org/doc/stable/reference/generated/numpy.i0.html

Snaking through the numpy source code I land at a C call and am struggling to drill down to the actual source code of that function.

There are some citations to math books published in the 60s but I don't have access to those.

One trail led to https://metacpan.org/dist/Math-Cephes/view/lib/Math/Cephes.pod#y0:-Bessel-function-of-the-second-kind,-order-zero which says this... basically it depends on `j0()`, another function we have to hunt down

```
DESCRIPTION:
Returns Bessel function of the second kind, of order
zero, of the argument.
 
The domain is divided into the intervals [0, 5] and
(5, infinity). In the first interval a rational approximation
R(x) is employed to compute
  y0(x)  = R(x)  +   2 * log(x) * j0(x) / PI.
Thus a call to j0() is required.
 
In the second interval, the Hankel asymptotic expansion
is employed with two rational functions of degree 6/6
and 7/7.
```

I'm probably going to put this down and close this issue because this is taking a lot of my time and I want to devote my firepower elsewhere. I appreciate your suggestion @ADD-eNavarro, and would love to get a Bessel filter added to FftSharp (and then it will be easy to add a Kaiser window). If you can come-up with a function that returns Bessel values with C#, I'd be happy to pick this up again.

Based on the Numpy docs I'd expect it to look like this:
```cs

```

```
```

**Comment by swharden** [comment]
... okay I kept looking and got a lead, so I'm opening this issue back up. [Accord.NET](https://github.com/accord-net/framework) is GPL-licensed so we have to be ***very careful*** not to let that code get into the FftSharp library or it would lose its ability to be released under the MIT license.

https://github.com/accord-net/framework/blob/49f2932a7baf69c7f6113ec82976aecca0c50353/Sources/Accord.Math/Functions/Bessel.cs#L376-L406

**Comment by swharden** [comment]
I found a repository that has a C++ Bessel function generator that is MIT licensed. Thanks @oygx210!

* https://github.com/oygx210/navguide
* https://github.com/oygx210/navguide/blob/cfae1c4b7b9228d009b6e7e71c3d6a1ee17462c0/src/common/bessel.c#L29-L68

**Comment by swharden** [comment]
What a roller coaster... but it's in now and published on NuGet (version 1.0.12) 

Let me know if it doesn't do the trick! Good luck with your project 👍 🚀

```cs
double[] window = FftSharp.Window.Kaiser(audio.Length, beta: 15);
```

![windows](https://user-images.githubusercontent.com/4165489/136313512-1df263e3-1c96-43e1-8661-1825c83a43ea.png)
